cpu/sam0_common/sam0_sdhc/sdhc.c

```diff
@@ -0,0 +1,1062 @@
+        volatile unsigned count = (us * CLOCK_CORECLOCK) / US_PER_SEC;
+        while (--count) {}
+    }
+}
+
+static void _reset_all(sdhc_state_t *state)
+{
+    SDHC_DEV->SRR.reg = SDHC_SRR_SWRSTALL;
+    while (SDHC_DEV->SRR.bit.SWRSTALL) {}
+    state->need_init = true;
+    state->error = 0;
+}
+
+static uint32_t _wait_for_event(sdhc_state_t *state)
+{
+    uint32_t res;
+
+    /* SDHC runs off CPU clock - block IDLE the clock does not stop */
+    pm_block(3);
+    mutex_lock(&state->sync);
+    pm_unblock(3);
+
+    res = state->error;
+    state->error = 0;
+    return res;
+}
+
+static void _init_clocks(sdhc_state_t *state)
+{
+    sam0_gclk_enable(SDHC_CLOCK);
+    sam0_gclk_enable(SDHC_CLOCK_SLOW);
+
+    gpio_init_mux(state->cd, SAM0_SDHC_MUX);
+    if (gpio_is_valid(state->wp)) {
+        gpio_init_mux(state->wp, SAM0_SDHC_MUX);
+    }
+
+    if (SDHC_DEV == SDHC0) {
+        /* data pins are fixed */
+        gpio_init_mux(SAM0_SDHC0_PIN_SDDAT0, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC0_PIN_SDDAT1, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC0_PIN_SDDAT2, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC0_PIN_SDDAT3, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC0_PIN_SDCMD, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC0_PIN_SDCK, SAM0_SDHC_MUX);
+
+        GCLK->PCHCTRL[SDHC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
+                                         | GCLK_PCHCTRL_GEN(SDHC_CLOCK);
+        GCLK->PCHCTRL[SDHC0_GCLK_ID_SLOW].reg = GCLK_PCHCTRL_CHEN
+                                              | GCLK_PCHCTRL_GEN(SDHC_CLOCK_SLOW);
+        MCLK->AHBMASK.bit.SDHC0_ = 1;
+        isr_ctx_0 = state;
+        NVIC_EnableIRQ(SDHC0_IRQn);
+    }
+
+#ifdef SDHC1
+    if (SDHC_DEV == SDHC1) {
+        /* data pins are fixed */
+        gpio_init_mux(SAM0_SDHC1_PIN_SDDAT0, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC1_PIN_SDDAT1, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC1_PIN_SDDAT2, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC1_PIN_SDDAT3, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC1_PIN_SDCMD, SAM0_SDHC_MUX);
+        gpio_init_mux(SAM0_SDHC1_PIN_SDCK, SAM0_SDHC_MUX);
+
+        GCLK->PCHCTRL[SDHC1_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
+                                         | GCLK_PCHCTRL_GEN(SDHC_CLOCK);
+        GCLK->PCHCTRL[SDHC1_GCLK_ID_SLOW].reg = GCLK_PCHCTRL_CHEN
+                                              | GCLK_PCHCTRL_GEN(SDHC_CLOCK_SLOW);
+        MCLK->AHBMASK.bit.SDHC1_ = 1;
+        isr_ctx_1 = state;
+        NVIC_EnableIRQ(SDHC1_IRQn);
+    }
+#endif
+}
+
+int sdhc_init(sdhc_state_t *state)
+{
+    bool f8;
+    uint32_t response;
+
+    /* set the initial clock slow, single bit and normal speed */
+    state->type = CARD_TYPE_SD;
+    state->version = CARD_VER_UNKNOWN;
+    state->rca = 0;
+    state->bus_width = 1;
+    state->high_speed = false;
+    state->clock = SDHC_SLOW_CLOCK_HZ;
+    mutex_init(&state->lock);
+    mutex_init(&state->sync);
+    mutex_lock(&state->sync);
```

Comment:
```suggestion
    const mutex_t _init_locked = MUTEX_INIT_LOCKED;
    state->sync = _init_locked;
```

Safes a function call.